Mathlib/AlgebraicGeometry/Nilpotents.lean

```diff
@@ -0,0 +1,213 @@
+/-
+Copyright (c) 2024 Christian Merten. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Christian Merten, Andrew Yang
+-/
+import Mathlib.AlgebraicGeometry.AffineScheme
+import Mathlib.AlgebraicGeometry.Cover.Open
+import Mathlib.AlgebraicGeometry.GammaSpecAdjunction
+import Mathlib.AlgebraicGeometry.Restrict
+
+/-!
+
+# Nilpotents and zero locus on schemes
+
+In this file we show some properties of the zero locus of a set of sections specific
+to locally ringed spaces or schemes and characterise nilpotent sections in terms of
+the associated basic open.
+
+In the case of affine schemes, the zero locus corresponds under the `Scheme.isoSpec` isomorphism
+to the zero locus in the sense of prime spectra. We provide API for reasoning about zero loci on
+affine schemes to avoid using the `Scheme.isoSpec` isomorphism in applications.
+
+## Main results
+
+- `Scheme.isNilpotent_of_basicOpen_eq_bot_of_isCompact`: a section over a compact open
+  of a scheme is nilpotent if and only if the associated basic open is empty.
+
+## TODO
+
+- Relate the zero locus to evaluation of sections in the residue fields of points.
+
+-/
+
+universe u
+
+open TopologicalSpace CategoryTheory Opposite
+
+noncomputable section
+
+namespace AlgebraicGeometry
+
+section toΓSpecFun
+
+/-- On an affine scheme, `toΓSpecFun` is bijective. -/
+lemma Scheme.toΓSpecFun_bijective_of_isAffine (X : Scheme.{u}) [IsAffine X] :
+    Function.Bijective X.toΓSpecFun :=
```

Comment:
You should use `X.isoSpec.hom.1.base` instead of this.

Reply:
Then for affine schemes the simp NF of `toΓSpecFun` is `isoSpec.hom.val.base`?

Reply:
You could say that, but essentially toΓSpecFun is a implementation detail and should not be used outside that file.

Reply:
In what terms do I state a lemma like this then?
```
lemma LocallyRingedSpace.toΓSpec_preim_zeroLocus_eq {X : LocallyRingedSpace}
    (s : Set (X.presheaf.obj (op ⊤))) :
    X.toΓSpecFun ⁻¹' PrimeSpectrum.zeroLocus s = X.toRingedSpace.zeroLocus s := by
```

Reply:
`(LocallyRingedSpace.toΓSpec X).1.base` I think. Please also provide it in terms of schemes anyway.

Reply:
Thanks. Maybe `toΓSpecFun` and friends should be private.

Reply:
Re in terms of schemes: What is the canonical spelling there for the unit of the adjunction for schemes? Do I still write `toΓSpec`?

Reply:
BTW: `(LocallyRingedSpace.toΓSpec X).1.base` simps to `toΓSpecBase`.

Reply:
Probably `ΓSpec.adjunction.unit.app X`

Reply:
Let me suggest the following: For now I do `ΓSpec.adjunction.unit.app X`, but I propose the following refactor:
- split up the gamma spec adjunction file
- make the constructions private
- add `Scheme.toΓSpec` and restate things in terms of that.
BTW: I think more files can be split, maybe we need some more folders and `Basic` files.